### PR TITLE
Add ExtractStatements to C API 

### LINKED ISSUE #4977: Please add ExtractStatements to C API
Hi,

we're using the C API in the Go DuckDB driver. All statements are processed using the `duckdb_prepare` function. If someone provides multiple statements to our `Exec` function, DuckDB throws an error, complaining about multiple statements in a prepare, which makes absolutely sense. See https://github.com/marcboeker/go-duckdb/issues/46

The Python driver uses the [`ExtractStatements` method](https://github.com/duckdb/duckdb/blob/master/tools/pythonpkg/src/pyconnection.cpp#L145) on the provided input and then prepares each extracted statement on its own.

Would it be possible to add the [`ExtractStatements` method](https://github.com/duckdb/duckdb/blob/master/src/include/duckdb/main/connection.hpp#L103) to the C API, so that we can also accept multiple statements in one call to our `Exec` function.

Thank you!

### PR DESCRIPTION
**Pull request by LindsayWray**
Addresses issue #4977

### CODE REVIEW

File: src/include/duckdb.h [hunk 1]
```diff
@@ -1007,6 +1008,48 @@ Executes the prepared statement with the given bound parameters, and returns an
...
+If the prepare fails, `duckdb_prepare_error` can be called to obtain the reason why the prepare failed.
+
+* connection: The connection object
+* extracted_statements: The extracted statements object
+* index: The index of the extracted statement to prepare
+* out_prepared_statement: The resulting prepared statement object
+* returns: `DuckDBSuccess` on success or `DuckDBError` on failure.
+*/
+duckdb_state duckdb_prepare_extracted_statement(duckdb_connection connection,
+                                                duckdb_extracted_statements extracted_statements, int index,
```
Comment (Mytherin) [inline src/include/duckdb.h:1044]: Should `index` be an `idx_t` instead?

File: src/include/duckdb.h [hunk 2]
```diff
@@ -1007,6 +1008,48 @@ Executes the prepared statement with the given bound parameters, and returns an
...
+/*!
+Extract all statements from a query.
+
+Note that after calling `duckdb_extract_statements`, the extracted statements should always be destroyed using
+`duckdb_destroy_extracted`, even if no statements were extracted.
+
+* connection: The connection object
+* query: The SQL query to extract
+* out_extracted_statements: The resulting extracted statements object
+* returns: The number of extracted statements or 0 on failure.
```
Comment (Mytherin) [inline src/include/duckdb.h:1025]: Perhaps we want some way of getting the error message as well?

File: test/api/capi/test_capi_extract.cpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+TEST_CASE("Test extract statements in C API", "[capi]") {
+	CAPITester tester;
+	duckdb_result res;
+	duckdb_extracted_statements stmts = nullptr;
+	duckdb_state status;
+
+	REQUIRE(tester.OpenDatabase(nullptr));
+
+	idx_t size = duckdb_extract_statements(tester.connection,
```
Comment (Mytherin) [inline test/api/capi/test_capi_extract.cpp:14]: Could we add a test with a parser failure in the `extract_statements`?

What about if the query is the empty string?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good. Some comments:

**Comment by LindsayWray** [comment]
I'll make a new pr, i'm having too much trouble with merge conflicts due to diffs in master and feature branch I think

**Pull request closed without merging**